external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2104,213 +2071,651 @@ fn add_field_types<T>(
+                    sp(dloc, Type_::UnresolvedError)
                 }
             };
-            let ty = sp(loc, Ref(mut_, desired_inner_ty));
-            T::exp(ty, sp(dloc, e_))
-        }
-        ExpDotted_::Dot(lhs, field, field_ty) => {
-            let lhs_borrow = exp_dotted_to_borrow(context, dloc, mut_, *lhs);
-            let sp!(tyloc, unfolded_) = core::unfold_type(&context.subst, lhs_borrow.ty.clone());
-            let lhs_mut = match unfolded_ {
-                Ref(lhs_mut, _) => lhs_mut,
-                _ => panic!(
-                    "ICE expected a ref from exp_dotted borrow, otherwise should have gotten a \
-                     TmpBorrow"
-                ),
+            inner.loc = dloc;
+            let accessor = ExpDottedAccess::Index {
+                index_loc: dloc,
+                syntax_methods,
+                args,
+                base_type,
             };
-            // lhs is immutable and current borrow is mutable
-            if !lhs_mut && mut_ {
-                context.env.add_diag(diag!(
-                    ReferenceSafety::RefTrans,
-                    (loc, "Invalid mutable borrow from an immutable reference"),
-                    (tyloc, "Immutable because of this position"),
-                ))
-            }
-            let e_ = TE::Borrow(mut_, Box::new(lhs_borrow), field);
-            let ty = sp(loc, Ref(mut_, field_ty));
-            T::exp(ty, sp(dloc, e_))
+            inner.accessors.push(accessor);
+            inner
         }
     }
 }
 
-fn exp_dotted_to_owned_value(
+fn exp_dotted_usage(
     context: &mut Context,
     usage: DottedUsage,
     eloc: Loc,
-    edot: ExpDotted,
-    inner_ty: Type,
-) -> T::Exp {
+    ndotted: N::ExpDotted,
+) -> Box<T::Exp> {
+    let constraint_verb = match &ndotted.value {
+        N::ExpDotted_::Exp(_) => None,
+        _ if matches!(usage, DottedUsage::Borrow(_)) => Some("borrow"),
+        N::ExpDotted_::Dot(_, _) => Some("dot access"),
+        N::ExpDotted_::Index(_, _) => Some("index"),
+    };
+    let edotted = process_exp_dotted(context, constraint_verb, ndotted);
+    if matches!(usage, DottedUsage::Borrow(_)) && edotted.accessors.is_empty() {
+        context.add_base_type_constraint(eloc, "Invalid borrow", edotted.base.ty.clone());
+    }
+    resolve_exp_dotted(context, usage, eloc, edotted)
+}
+
+fn exp_dotted_expression(
+    context: &mut Context,
+    usage: DottedUsage,
+    constraint_verb: Option<&str>,
+    eloc: Loc,
+    ndotted: N::ExpDotted,
+) -> Box<T::Exp> {
+    let edotted = process_exp_dotted(context, constraint_verb, ndotted);
+    if matches!(usage, DottedUsage::Borrow(_)) && edotted.accessors.is_empty() {
+        context.add_base_type_constraint(eloc, "Invalid borrow", edotted.base.ty.clone());
+    }
+    resolve_exp_dotted(context, usage, eloc, edotted)
+}
+
+// This comment servees to document the function below. Depending on the shape of the dotted
```

Comment:
Everything below here is a thing of beauty. Sincerely, thanks for cleaning this up, its been piled upon for long enough